### PR TITLE
WT-8497 Initial riscv64 support

### PR DESCRIPTION
**Pull request by acmorrow**

### CODE REVIEW

**Comment by acmorrow** [comment]
A few notes on this:
- Much of the build system changes were made via copy/paste with mutation by analogy. I was able to get autoconf to run and build in a riscv64 QEMU environment and run some tests, so I think at least the autotools build is OK.
- The CMake changes are untested
- I added a crc32 thing for riscv64 because it seemed to be needed. Maybe it would be better if there were just a generic fallback to the `sw` version if there isn't  hardware specialization? LMK if you want me to pursue that instead of adding a non-accelerated version for riscv64.

I'm also going to leave a few more detailed notes on the changes themselves, where interesting.

File: SConstruct
Comment (acmorrow) [inline SConstruct:256]: If `SConstruct` is really only for Windows builds, this could probably be dropped, but I did see that there was support here for the other architectures.

File: build_posix/configure.ac.in
Comment (acmorrow) [inline build_posix/configure.ac.in:67]: Pure copy paste from the nearby arm stanza, but I did an autoreconf and configure / make and it all seemed to work OK.

File: cmake/configs/base.cmake
Comment (acmorrow) [inline cmake/configs/base.cmake:21]: By analogy with WT_AARCH64 case.

File: cmake/configs/riscv64/linux/config.cmake
Comment (acmorrow) [inline cmake/configs/riscv64/linux/config.cmake:1]: By analogy with the aarch64 one.

File: cmake/configs/riscv64/linux/config.cmake
Comment (acmorrow) [inline cmake/configs/riscv64/linux/config.cmake:23]: See https://www.sifive.com/blog/all-aboard-part-1-compiler-args for background on the `mafdc` and `lp64d` arguments here.

File: cmake/helpers.cmake
Comment (acmorrow) [inline cmake/helpers.cmake:613]: By analogy with aarch64

File: cmake/toolchains/riscv64/linux/plat_clang.cmake
Comment (acmorrow) [inline cmake/toolchains/riscv64/linux/plat_clang.cmake:10]: By analogy, untested

File: cmake/toolchains/riscv64/linux/plat_gcc.cmake
Comment (acmorrow) [inline cmake/toolchains/riscv64/linux/plat_gcc.cmake:10]: By analogy, untested

File: dist/filelist
Comment (acmorrow) [inline dist/filelist:53]: I'd have preferred not to need to add this since we don't actually have hw acceleration on this platform (yet), but I couldn't figure out how to instruct it to do a graceful fallback to sw without adding such a file. Please let me know if there is another way.

File: dist/prototypes.py
Comment (acmorrow) [inline dist/prototypes.py:63]: By analogy, untested.

File: src/checksum/riscv64/crc32-riscv64.c
Comment (acmorrow) [inline src/checksum/riscv64/crc32-riscv64.c:46]: The minimal implementation that made things work. I'd be happy to not need this file, per the above.

File: src/include/gcc.h
Comment (acmorrow) [inline src/include/gcc.h:285]: There is a `pause` instruction in RISC-V but I don't think compilers support it yet: https://riscv.org/announcements/2021/02/risc-v-international-unveils-fast-track-architecture-extension-process-and-ratifies-zihintpause-extension/

It is interesting, because it is a `fence` instruction with a null successor set. I didn't want to gamble on writing it that way, so I just went with `nop` for now.

File: src/include/gcc.h
Comment (acmorrow) [inline src/include/gcc.h:286]: This file lacks a detailed explanation of what the intended semantics of `WT_FULL_BARRIER`, `WT_READ_BARRIER`, and `WT_WRITE_BARRIER` really are.

However, since the x86_64 mapping is:
- `FULL` : `mfence`
- `READ`: `lfence`
- `WRITE`: `sfence`

My understanding is that `lfence` and `sfence` only synchronize with other `[lm]fence` and `[sm]fence` instructions. Reading https://five-embeddev.com/riscv-isa-manual/latest/memory.html#sec:mm:fence, it seems then that on riscv where we can specify the predecessor set and successor set for the fence, that we should have:
- `FULL`: `rw, rw`
- `READ`: `r, r`
- `WRITE`: `w, w`

I'd like to dig a little deeper to make sure that these are the right `fence` intsructions in general, but it seems plausible.

**Comment by acmorrow** [comment]
One real concern with this of course is testing. I don't have a great answer there. I did the development work for this in the mongodb tree where I used SCons to cross compile to riscv64 on Ubuntu, and then I did runtime testing with a `qemu-debootstrap`'d riscv `schroot`. That's enough to show it builds and runs, to some extent, but I'm not sure if that meets the bar for commit. In any event, I wanted to share this work.

File: SConstruct
Comment (keithbostic) [reply SConstruct:256]: Yes; SConstruct does have some limited support for platforms other than Windows, probably because at the time, there was an expectation we'd switch to SCons as the WiredTiger build platform. Since CMake was chosen, I don't know what the deal is, but my guess is that the right approach is to strip everything not related to Windows from SConstruct.  @alisonfel, you know more about this than I do, so I'll defer to you on whether this needs a cleanup ticket.

File: dist/filelist
Comment (keithbostic) [reply dist/filelist:53]: For lots of reasons (including that MongoDBb builds out the checksum code as a separate, non-WiredTiger library and gcc and clang have their own weird syntax requirements), this is just a configuration mess. Having a separate file is probably the cleanest thing to do and doesn't cost us any performance, I think it's the right choice and avoids file shuffling when we add acceleration.

File: src/include/gcc.h
Comment (keithbostic) [reply src/include/gcc.h:285]: I'd suggest a comment here, plus links?

File: src/include/gcc.h
Comment (keithbostic) [reply src/include/gcc.h:286]: Agree, this makes sense to me.

**Review by keithbostic — APPROVED** [review]
Thanks, Andrew!

@alisonfel, I ignored all of the cmake changes.

If you want something to stress your barriers a bit, you can run `test/format` on a standalone build, also there are some heavily threaded programs in the standalone tests that will fail if there are barrier problems.

**Comment by etienneptl** [comment]
@acmorrow, if you run locally `s_all` from the `dist` folder, you should see why some of the PR tests fail.

The only error I can see so far is:
```
[2021/12/06 22:12:30.264] Update src/docs/spell.ok to remove warnings.
[2021/12/06 22:12:30.291] 	riscv
```

You will need to update the `spell.ok` file from the `docs` folder with the missing word `riscv`.

File: src/include/gcc.h
Comment (acmorrow) [reply src/include/gcc.h:285]: Added, will push an update soon.

File: src/include/gcc.h
Comment (acmorrow) [reply src/include/gcc.h:286]: Actually, I found that the file does have a good explanation, I just skimmed past it. Based on the description, I'm still more confident that `r, r` correctly models `READ` and `w, w` is correct for `WRITE`. I've added a comment with a link, it will show up on the next push.

**Comment by acmorrow** [comment]
> If you want something to stress your barriers a bit, you can run test/format on a standalone build, also there are some heavily threaded programs in the standalone tests that will fail if there are barrier problems.

@keithbostic - Since my testing platform is `qemu` I'm not sure how the fence instructions get mapped, so I'm not sure that testing would reveal much. I've ordered some RISC-V hardware, so in a couple months I'll be able to run those tests in a more meaningful environment.

**Comment by acmorrow** [comment]
@keithbostic, @etienneptl - Updated, PTAL.

@alisonfel  - Would love your thoughts on build system stuff.

File: SConstruct
Comment (keithbostic) [reply SConstruct:256]: There's also the whole "build checksum stuff as a separate library", which I've kind of lost track of, but yeah, I'm good with whatever @alisonfel decides.

File: src/checksum/riscv64/crc32-riscv64.c
Comment (keithbostic) [reply src/checksum/riscv64/crc32-riscv64.c:46]: We've all agreed to keep the additional file.

**Comment by keithbostic** [comment]
@acmorrow, the PR system runs basic scripting over the source tree, and some of the checks are failing. If you run `dist/s_all`, you should see the same failures locally.

**Comment by acmorrow** [comment]
@keithbostic - Yeah, I just did something silly updating the spelling file. I've pushed an update that I think will fix those items.

**Comment by acmorrow** [comment]
@keithbostic - I don't know how to get `s_all` to pass. The `s_clang-format` seems to be unstable: each run wants to add a new blank line after the long URL int he comment block. Ditto `s_longlines` complaining about the long line. Could you maybe pull this locally and see if you can get it to work for you?

**Comment by keithbostic** [comment]
> @keithbostic - I don't know how to get `s_all` to pass. The `s_clang-format` seems to be unstable: each run wants to add a new blank line after the long URL int he comment block. Ditto `s_longlines` complaining about the long line. Could you maybe pull this locally and see if you can get it to work for you?

Sure, will do.

File: cmake/configs/riscv64/linux/config.cmake
Comment (sauclovian-wt) [inline cmake/configs/riscv64/linux/config.cmake:22]: You might want to update this comment, which seems to have been pasted from the arm file :-)

**Comment by sauclovian-wt** [comment]
All seems ok to me; don't see any reason the cmake stuff shouldn't work, but I'm not a cmake expert.

File: src/include/gcc.h
Comment (sauclovian-wt) [reply src/include/gcc.h:296]: I can't get the assembler I have to accept any likely syntax, though. "fence w,0" is rejected and "fence w," assembles to "fence w,w". Blah :-(

**Review by alisonfel — APPROVED** [review]
Hey @acmorrow, thanks for these changes. 
I had a look at the CMake changes and they all make sense.. Using the other architecture configs as an analogy is more or less the correct approach here :) 
This LGTM. I have a comment below, but I don't want it to be blocking (as is it could be an issue on my end)

***

I wanted to ask if cross-compiling with the gcc toolchain has worked on your end? I gave it a go and wound up with linker errors, having undefined symbols for the atomic functions:
```
libwiredtiger.so.10.0.1: undefined reference to `__atomic_compare_exchange_2'
libwiredtiger.so.10.0.1: undefined reference to `__atomic_compare_exchange_1'
```
This does seem a little odd, I would've thought these were builtins. Manually pass through a linker flag to the atomics library fixes the issue `-latomic`.
The cmake command I ran: ` cmake -DCMAKE_TOOLCHAIN_FILE=../cmake/toolchains/gcc.cmake -DCMAKE_CROSSCOMPILING=TRUE -DWT_ARCH=riscv64 -DWT_OS=linux -DCMAKE_PREFIX_PATH=/usr/riscv64-linux-gnu/lib -G Ninja ../.`
(`riscv64-linux-gnu-gcc -v` -> `gcc version 7.5.0`)

I don't want this to block the PR on this as cross-compiling is not a major use case (its non-perfect for PPC and ZSeries also). This could easily be an issue with the toolchain setup on my end (as cross-compiling setups can be a bit finicky). 
More so curious if you encountered this? I'm guessing this is a non-issue when compiling on an actual (or emulated) RISCV host?

File: cmake/configs/riscv64/linux/config.cmake
Comment (acmorrow) [reply cmake/configs/riscv64/linux/config.cmake:22]: Done

Comment (acmorrow) [reply cmake/configs/riscv64/linux/config.cmake:22]: Done

File: src/include/gcc.h
Comment (acmorrow) [reply src/include/gcc.h:296]: Thanks for trying it out though!

**Comment by acmorrow** [comment]
> I wanted to ask if cross-compiling with the gcc toolchain has worked on your end? I gave it a go and wound up with linker errors, having undefined symbols for the atomic functions:

@alisonfel - I did cross compile, but I did so via the server's SCons-driven build of WT (the fourth build system!). The fact that you needed to manually add `-latomic` isn't very surprising to me, you do sometimes need that. I think the server build may add it automatically when required, which was why I didn't need to do so.

**Comment by acmorrow** [comment]
Thanks all for the review, I think this is ready to merge. What is the process?

**Comment by keithbostic** [comment]
> Thanks all for the review, I think this is ready to merge. What is the process?

I'll merge it as soon as evergreen confirms.

**Pull request merged**
